cdap-api-common/src/main/java/co/cask/cdap/api/data/format/StructuredRecord.java

```diff
@@ -71,6 +79,174 @@ public Schema getSchema() {
+    if (value == null) {
+      return null;
+    }
+
+    if (logicalType == LogicalType.TIMESTAMP_MILLIS) {
+      return getZonedDateTime((long) value, zoneId, TimeUnit.MILLISECONDS);
+    }
+
+    return getZonedDateTime((long) value, zoneId, TimeUnit.MICROSECONDS);
+  }
+
+  /**
+   * Get zoned date and time represented by the field.
+   *
+   * @param zoneId    zone id for the field
+   * @return {@link ZonedDateTime} represented by field.
+   */
+  private ZonedDateTime getZonedDateTime(long ts, ZoneId zoneId, TimeUnit unit) {
+    // get fractional portion of timestamp, we always store timestamp in micro seconds, so first get that portion.
+    // for example: timestamp is 1970-01-01T00:00:01.112456Z, here fraction = 112456, tsInSeconds = 1
+    long divisor = unit == TimeUnit.MICROSECONDS ? 1000000L : 1000L;
+    int fraction = (int) (ts % divisor);
+    long tsInSeconds = unit.toSeconds(ts);
+    // create an Instant with time in seconds and fraction which will be stored as nano seconds.
+    Instant instant = Instant.ofEpochSecond(tsInSeconds, unit.toNanos(fraction));
+    return ZonedDateTime.ofInstant(instant, zoneId);
+  }
+
+  /**
+   * Validate and get underlying logical type if the schema is a union type.
+   *
+   * @param fieldName    name of the field
+   * @param field        field with logical type
+   * @param logicalTypes acceptable logical types
+   * @return {@link LogicalType} underlying logical type that field represents
+   * @throws IllegalArgumentException if the field name is not found in the map of fields or the field provided
+   *                                  is not an acceptable type
+   */
+  private static LogicalType validateAndGetLogicalType(String fieldName, Schema.Field field,
+                                                       EnumSet<LogicalType> logicalTypes) {
+    if (field == null) {
+      throw new IllegalArgumentException(String.format("Field %s provided does not exist in the schema.", fieldName));
+    }
+
+    Schema fieldSchema = field.getSchema();
+
+    if (fieldSchema.getType() != Schema.Type.UNION) {
+      if (!fieldSchema.getType().isSimpleType() || fieldSchema.getLogicalType() == null) {
+        throw new IllegalArgumentException(String.format("Field %s does not have a logical type.", fieldName));
+      }
+
+      if (!logicalTypes.contains(fieldSchema.getLogicalType())) {
+        throw new IllegalArgumentException(String.format("Field %s must be of logical type %s, " +
+                                                           "instead it is of type %s",
+                                                         fieldName, logicalTypes,
+                                                         fieldSchema.getLogicalType()));
+      }
+    }
+
+    LogicalType parsedType = getLogicalTypeFromUnionSchema(fieldSchema, logicalTypes);
+    if (parsedType != null) {
+      return parsedType;
+    }
+
+    return fieldSchema.getLogicalType();
+  }
```

Comment:
The newly added methods would be cleaner if you do this:

```java
  /**
   * Validates and returns the underlying {@link LogicalType} for the given field.
   *
   * @param fieldName name of the field
   * @param allowedTypes accetable logical types
   * @return the underlying {@link LogicalType} for the field
   * @throws IllegalArgumentException if the field name is not found in the map of fields or the field provided
   *                                  is not an acceptable type
   */
  private LogicalType validateAndGetLogicalType(String fieldName, Set<LogicalType> allowedTypes) {
    //
  }

  /**
   * Validates and returns the underlying {@link LogicalType} of the given {@link Schema.Field}.
   *
   * @param field field with logical type
   * @param allowedTypes acceptable logical types
   * @return the underlying {@link LogicalType} for the field
   * @throws IllegalArgumentException if the field provided does not have a {@link LogicalType} that is
   *                                  one of the acceptable types
   */
  private static LogicalType validateAndGetLogicalType(Schema.Field field, Set<LogicalType> allowedTypes) {
    //
  }

  /**
   * Gets the {@link LogicalType} defined in the given {@link Schema}.
   * If the given {@link Schema} is a {@link Schema.Type#UNION}, this method will find
   * the {@link LogicalType} recursively and return the first {@link LogicalType} it encountered.
   *
   * @param schema the {@link Schema} for finding the {@link LogicalType}
   * @return the {@link LogicalType} defined in the given schema or {@code null} if no {@link LogicalType} was found
   */
  @Nullable
  private static LogicalType getLogicalType(Schema schema) {
    ///
  }
```

Here is the complete diff of how I think should look like:

```
diff --git a/cdap-api-common/src/main/java/co/cask/cdap/api/data/format/StructuredRecord.java b/cdap-api-common/src/main/java/co/cask/cdap/api/data/format/StructuredRecord.java
index cb8af40de2..bd31877563 100644
--- a/cdap-api-common/src/main/java/co/cask/cdap/api/data/format/StructuredRecord.java
+++ b/cdap-api-common/src/main/java/co/cask/cdap/api/data/format/StructuredRecord.java
@@ -34,6 +34,7 @@ import java.util.EnumSet;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Objects;
+import java.util.Set;
 import java.util.TimeZone;
 import java.util.concurrent.TimeUnit;
 import javax.annotation.Nullable;
@@ -89,7 +90,7 @@ public class StructuredRecord implements Serializable {
    */
   @Nullable
   public LocalDate getDate(String fieldName) {
-    validateAndGetLogicalType(fieldName, schema.getField(fieldName), EnumSet.of(LogicalType.DATE));
+    validateAndGetLogicalType(fieldName, EnumSet.of(LogicalType.DATE));
     Object value = fields.get(fieldName);
     return value == null ? null : LocalDate.ofEpochDay(((Integer) value).longValue());
   }
@@ -105,9 +106,8 @@ public class StructuredRecord implements Serializable {
    */
   @Nullable
   public LocalTime getTime(String fieldName) {
-    LogicalType logicalType =  validateAndGetLogicalType(fieldName, schema.getField(fieldName),
-                                                                EnumSet.of(LogicalType.TIME_MILLIS,
-                                                                           LogicalType.TIME_MICROS));
+    LogicalType logicalType = validateAndGetLogicalType(fieldName, EnumSet.of(LogicalType.TIME_MILLIS,
+                                                                              LogicalType.TIME_MICROS));
     Object value = fields.get(fieldName);
     if (value == null) {
       return null;
@@ -147,9 +147,8 @@ public class StructuredRecord implements Serializable {
    */
   @Nullable
   public ZonedDateTime getTimestamp(String fieldName, ZoneId zoneId) {
-    LogicalType logicalType = validateAndGetLogicalType(fieldName, schema.getField(fieldName),
-                                                               EnumSet.of(LogicalType.TIMESTAMP_MILLIS,
-                                                                          LogicalType.TIMESTAMP_MICROS));
+    LogicalType logicalType = validateAndGetLogicalType(fieldName, EnumSet.of(LogicalType.TIMESTAMP_MILLIS,
+                                                                              LogicalType.TIMESTAMP_MICROS));
     Object value = fields.get(fieldName);
     if (value == null) {
       return null;
@@ -179,71 +178,65 @@ public class StructuredRecord implements Serializable {
     return ZonedDateTime.ofInstant(instant, zoneId);
   }

+  private LogicalType validateAndGetLogicalType(String fieldName, Set<LogicalType> allowedTypes) {
+    Schema.Field field = schema.getField(fieldName);
+    if (field == null) {
+      throw new IllegalArgumentException("Field " + fieldName + " does not exists in the schema");
+    }
+    return validateAndGetLogicalType(field, allowedTypes);
+  }
+
   /**
-   * Validate and get underlying logical type if the schema is a union type.
+   * Validates and returns the underlying {@link LogicalType} of the given {@link Schema.Field}.
    *
-   * @param fieldName    name of the field
-   * @param field        field with logical type
-   * @param logicalTypes acceptable logical types
-   * @return {@link LogicalType} underlying logical type that field represents
+   * @param field field with logical type
+   * @param allowedTypes acceptable logical types
+   * @return the underlying {@link LogicalType} for the field
    * @throws IllegalArgumentException if the field name is not found in the map of fields or the field provided
    *                                  is not an acceptable type
    */
-  private static LogicalType validateAndGetLogicalType(String fieldName, Schema.Field field,
-                                                       EnumSet<LogicalType> logicalTypes) {
+  private static LogicalType validateAndGetLogicalType(Schema.Field field, Set<LogicalType> allowedTypes) {
     if (field == null) {
-      throw new IllegalArgumentException(String.format("Field %s provided does not exist in the schema.", fieldName));
+      return null;
     }

-    Schema fieldSchema = field.getSchema();
-
-    if (fieldSchema.getType() != Schema.Type.UNION) {
-      if (!fieldSchema.getType().isSimpleType() || fieldSchema.getLogicalType() == null) {
-        throw new IllegalArgumentException(String.format("Field %s does not have a logical type.", fieldName));
-      }
-
-      if (!logicalTypes.contains(fieldSchema.getLogicalType())) {
-        throw new IllegalArgumentException(String.format("Field %s must be of logical type %s, " +
-                                                           "instead it is of type %s",
-                                                         fieldName, logicalTypes,
-                                                         fieldSchema.getLogicalType()));
-      }
+    String fieldName = field.getName();
+    LogicalType logicalType = getLogicalType(field.getSchema());
+    if (logicalType == null) {
+      throw new IllegalArgumentException(String.format("Field %s does not have a logical type.", fieldName));
     }
-
-    LogicalType parsedType = getLogicalTypeFromUnionSchema(fieldSchema, logicalTypes);
-    if (parsedType != null) {
-      return parsedType;
+    if (!allowedTypes.contains(logicalType)) {
+      throw new IllegalArgumentException(
+        String.format("Field %s must be of logical type %s, instead it is of type %s",
+                      fieldName, allowedTypes, logicalType));
     }

-    return fieldSchema.getLogicalType();
+    return logicalType;
   }

   /**
-   * Parse through logical type and get underlying {@link LogicalType}
-   * @param fieldSchema name of the field
-   * @param logicalTypes acceptable logical types
-   * @return {@link LogicalType} represented by the union schema
+   * Gets the {@link LogicalType} defined in the given {@link Schema}.
+   * If the given {@link Schema} is a {@link Schema.Type#UNION}, this method will find
+   * the {@link LogicalType} recursively and return the first {@link LogicalType} it encountered.
+   *
+   * @param schema the {@link Schema} for finding the {@link LogicalType}
+   * @return the {@link LogicalType} defined in the given schema or {@code null} if no {@link LogicalType} was found
    */
-  private static LogicalType getLogicalTypeFromUnionSchema(Schema fieldSchema, EnumSet<LogicalType> logicalTypes) {
-    if (fieldSchema.getType().isSimpleType()) {
-      if (fieldSchema.getLogicalType() == null || !logicalTypes.contains(fieldSchema.getLogicalType())) {
-        return null;
-      }
+  @Nullable
+  private static LogicalType getLogicalType(Schema schema) {
+    LogicalType logicalType = schema.getLogicalType();
+    if (logicalType != null) {
+      return logicalType;
     }

-    if (fieldSchema.getType() == Schema.Type.UNION) {
-      for (Schema unionSchema : fieldSchema.getUnionSchemas()) {
-        if (unionSchema.getLogicalType() != null && logicalTypes.contains(unionSchema.getLogicalType())) {
-          return unionSchema.getLogicalType();
-        }
-
-        LogicalType logicalType = getLogicalTypeFromUnionSchema(unionSchema, logicalTypes);
+    if (schema.getType() == Schema.Type.UNION) {
+      for (Schema unionSchema : schema.getUnionSchemas()) {
+        logicalType = getLogicalType(unionSchema);
         if (logicalType != null) {
           return logicalType;
         }
       }
     }
-
     return null;
   }

@@ -284,7 +277,7 @@ public class StructuredRecord implements Serializable {
      *                                   value is given
      */
     public Builder set(String fieldName, @Nullable Object value) {
-      createField(fieldName, value);
+      validateAndGetField(fieldName, value);
       fields.put(fieldName, value);
       return this;
     }
@@ -300,8 +293,7 @@ public class StructuredRecord implements Serializable {
      * @throws IllegalArgumentException if the provided date is an invalid date
      */
     public Builder setDate(String fieldName, @Nullable LocalDate localDate) {
-      Schema.Field field = createField(fieldName, localDate);
-      validateAndGetLogicalType(fieldName, field, EnumSet.of(LogicalType.DATE));
+      validateAndGetLogicalType(validateAndGetField(fieldName, localDate), EnumSet.of(LogicalType.DATE));
       if (localDate == null) {
         fields.put(fieldName, null);
         return this;
@@ -326,9 +318,9 @@ public class StructuredRecord implements Serializable {
      *                                   value is given
      */
     public Builder setTime(String fieldName, @Nullable LocalTime localTime) {
-      Schema.Field field = createField(fieldName, localTime);
-      LogicalType logicalType = validateAndGetLogicalType(fieldName, field, EnumSet.of(LogicalType.TIME_MILLIS,
-                                                                                       LogicalType.TIME_MICROS));
+      Schema.Field field = validateAndGetField(fieldName, localTime);
+      LogicalType logicalType = validateAndGetLogicalType(field, EnumSet.of(LogicalType.TIME_MILLIS,
+                                                                            LogicalType.TIME_MICROS));

       if (localTime == null) {
         fields.put(fieldName, null);
@@ -360,9 +352,9 @@ public class StructuredRecord implements Serializable {
      * @throws IllegalArgumentException if the provided date is an invalid timestamp
      */
     public Builder setTimestamp(String fieldName, @Nullable ZonedDateTime zonedDateTime) {
-      Schema.Field field = createField(fieldName, zonedDateTime);
-      LogicalType logicalType = validateAndGetLogicalType(fieldName, field, EnumSet.of(LogicalType.TIMESTAMP_MILLIS,
-                                                                                       LogicalType.TIMESTAMP_MICROS));
+      Schema.Field field = validateAndGetField(fieldName, zonedDateTime);
+      LogicalType logicalType = validateAndGetLogicalType(field, EnumSet.of(LogicalType.TIMESTAMP_MILLIS,
+                                                                            LogicalType.TIMESTAMP_MICROS));

       if (zonedDateTime == null) {
         fields.put(fieldName, null);
@@ -424,7 +416,7 @@ public class StructuredRecord implements Serializable {
     @Deprecated
     public Builder convertAndSet(String fieldName, @Nullable Date date,
                                  @Nullable DateFormat dateFormat) throws UnexpectedFormatException {
-      Schema.Field field = createField(fieldName, date);
+      Schema.Field field = validateAndGetField(fieldName, date);
       boolean isNullable = field.getSchema().isNullable();
       if (isNullable && date == null) {
         fields.put(fieldName, null);
@@ -454,7 +446,7 @@ public class StructuredRecord implements Serializable {
      *                                   value is given, or the string cannot be converted to the type for the field
      */
     public Builder convertAndSet(String fieldName, @Nullable String strVal) throws UnexpectedFormatException {
-      Schema.Field field = createField(fieldName, strVal);
+      Schema.Field field = validateAndGetField(fieldName, strVal);
       fields.put(fieldName, convertString(field.getSchema(), strVal));
       return this;
     }
@@ -521,7 +513,7 @@ public class StructuredRecord implements Serializable {
       }
     }

-    private Schema.Field createField(String fieldName, Object val) {
+    private Schema.Field validateAndGetField(String fieldName, Object val) {
       Schema.Field field = schema.getField(fieldName);
       if (field == null) {
         throw new UnexpectedFormatException("field " + fieldName + " is not in the schema.");
```